### DISCUSSION TITLE
retrieve camera position from interactive 3D interface

### DISCUSSION THREAD

**Discussion by ifermion** [pyvista/pyvista#3646] [Q&A] [answered]
Hello,

First of all, thanks a lot to all the developers for making available such a versatile library for making 3D visualisations in python.

I know that we can predefine the camera position and the focus for a Plotter object, so that when it's viewed the camera is placed at the coordinates already defined.

For example,

```
plotter = pv.Plotter(window_size=(800, 800)) 
plotter.camera.position = np.array([-7.,5.,1.])
plotter.show()
```

My question is whether it would be possible to retrieve the camera positions as we interact with the visual, as in rotate it and look at the scene in different directions. Is there any way to do a manual rotation once ```plotter.show()``` is called and know what the updated camera position is ? and similarly collect the camera positions at several view points without having to close the visualisation window ?

Thank you in advance !

**Comment by MatthewFlamm** [comment, marked as answer]
One thought is to print the coordinates (or something more sophisticated) using a key press event  https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.add_key_event.html.  You could bind it to the 'c' key for example and then gather the positions as often as you like.

**Reply by banesullivan** [reply]
+1. Marking as the solution and providing some example code:

```py
import pyvista as pv

class CameraTracker:
    def __init__(self, plotter):
        self.positions = []
        self.plotter = plotter
        
        # Add key event to plotter
        self.plotter.add_key_event('c', self)
        
    def __call__(self):
        self.positions.append(self.plotter.camera.position)

mesh = pv.Cone()

p = pv.Plotter(notebook=False)
tracker = CameraTracker(p)
p.add_mesh(mesh)
p.show()
```

```py
>>> tracker.positions
[(1.7713431113390736, 1.8176267517777245, 1.9549210197530709),
 (1.8857059562244372, -2.5748406770728307, -0.2782950856368634)]
```

**Reply by ifermion** [reply]
Thank you very much for both MatthewFlamm and banesullivan. I'll check this out. I'm sorry for the late reply. Have been busy with many things and hope to be able to test this soon.

**Reply by MatthewFlamm** [reply]
> +1. Marking as the solution and providing some example code:

Very nice example.  More useful than just printing the positions.

**Reply by ifermion** [reply]
Thanks again guys. I finally tested this and found this example very useful. I really liked the way you've achieved this by means of creating an object.
